### PR TITLE
[KEYCLOAK-18426] - Support required by role and scopes in Admin UI

### PR DESCRIPTION
**Pull request by pedroigor**

### CODE REVIEW

**Review by velias — DISMISSED** [review]

**Review by mposolda — DISMISSED** [review]
Approving, mostly based on the @velias review.

**Comment by velias** [comment]
Tested latest changes, Admin UI looks good to me. 
I'm only not sure about the automatic change of the user profile config when user removes scope, as it may change behavior of the profile significantly (some attribute will be selected or required for all cases after scope is removed) and admin may not realize this when he removes client scope.  Admin should be warned during the scope removal that user profile will be modified. Or user profile should not be modified so behavior doesn't change (attribute will never be required or selected as scope will never be requested) and admin will realize problem next time when profile config will be modified thanks to scope validation you added (which is good to have it I think)

File: services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java [hunk 1]
```diff
@@ -127,6 +130,25 @@ private static void validate(KeycloakSession session, UPAttribute attributeConfi
...
+        if (scopes == null) {
+            return;
+        }
+
+        for (String scope : scopes) {
+            RealmModel realm = session.getContext().getRealm();
+            Stream<ClientScopeModel> realmScopes = realm.getClientScopesStream();
+
+            if (!realmScopes.anyMatch(cs -> cs.getName().equals(scope))) {
+                errors.add(new StringBuilder("Invalid scope ").append(scope).toString());
```
Comment (velias) [inline services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java:150]: Error message should be probably better, like in role validtion, to name exact attribute and field where the problem is

File: services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java [hunk 2]
```diff
@@ -127,6 +130,25 @@ private static void validate(KeycloakSession session, UPAttribute attributeConfi
...
+        if (scopes == null) {
+            return;
+        }
+
+        for (String scope : scopes) {
+            RealmModel realm = session.getContext().getRealm();
+            Stream<ClientScopeModel> realmScopes = realm.getClientScopesStream();
+
+            if (!realmScopes.anyMatch(cs -> cs.getName().equals(scope))) {
+                errors.add(new StringBuilder("'").append(propertyName).append("' configuration for attribute '").append(attributeName).append("' contains unsupported role '").append(scope).append("'").toString());
```
Comment (velias) [inline services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java:150]: Ooops, looks like typo there `contains unsupported role` should be `contains unsupported scope`

Comment (pedroigor) [reply services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java]: Fixed

**Review by velias — DISMISSED** [review]

**Pull request closed without merging**